code/core/src/preview-api/modules/preview-web/PreviewWithSelection.tsx

```diff
@@ -41,8 +41,38 @@ import { StoryRender } from './render/StoryRender';
 const globalWindow = globalThis;
 
 function focusInInput(event: Event) {
-  const target = ((event.composedPath && event.composedPath()[0]) || event.target) as Element;
-  return /input|textarea/i.test(target.tagName) || target.getAttribute('contenteditable') !== null;
+  type EventWithComposedPath = Event & { composedPath: () => EventTarget[] };
+  const hasComposedPath = (value: Event): value is EventWithComposedPath =>
+    typeof (value as { composedPath?: unknown }).composedPath === 'function';
+
+  type ElementLike = Pick<Element, 'tagName' | 'getAttribute'>;
+  const isElementLike = (value: unknown): value is ElementLike => {
+    return (
+      value != null &&
+      typeof (value as { tagName?: unknown }).tagName === 'string' &&
+      typeof (value as { getAttribute?: unknown }).getAttribute === 'function'
+    );
+  };
+
+  const isEditableElement = (element: ElementLike) => {
+    if (/input|textarea/i.test(element.tagName)) {
+      return true;
+    }
+
+    if (element.getAttribute('contenteditable') !== null) {
+      return true;
+    }
+
+    // Closed shadow roots can retarget events to the custom element host.
+    // Allow web components to opt in to "editable" semantics via ARIA roles on the custom element host.
+    const role = element.getAttribute('role')?.toLowerCase();
+    return role === 'input' || role === 'searchbox' || role === 'textbox' || role === 'combobox';
+  };
+
+  const composedPath = hasComposedPath(event) ? event.composedPath() : [];
+  const candidates = composedPath.length > 0 ? composedPath : [event.target];
+
+  return candidates.some((candidate) => isElementLike(candidate) && isEditableElement(candidate));
```

Comment:
I don't see how this fixes anything for closed root WCs.

You still cannot inspect the element actually receiving focus inside the WC (whether it is an appropriate tag name, implicit role, explicit role or contenteditable). So instead, what I understand you did is that you asked WC authors, by convention, to add a `role` to the root of their closed shadow WC.

If my understanding is correct, doing so will result in ARIA violations as users are now misadvertising a role on a root element, when the actual interactive element is deeper inside the shadow DOM.

Unfortunately that second solution described on the issue was not correct. I will comment on the issue with alternative solutions.